### PR TITLE
Add a generic data type converter to the `Cursor` object

### PR DESCRIPTION
**Pull request by amotl**
Hi there,

### Introduction
After converging #395 to #437, it became clear / emerged that we wanted to have the type conversion

- only being optionally enabled with a feature flag /cc @mfussenegger 
- being implemented more elegantly /cc @matriv 

This patch aims to resolve both aspects.

### Details
This will allow converting fetched data from CrateDB data types to Python data types in different ways. Its usage is completely optional. When not used, the feature will not incur any overhead.

There is also a `DefaultTypeConverter`, which aims to become a sane default choice when looking at this kind of convenience. It will enable the user to work with native Python data types from the start, for all database types where this makes sense, without needing to establish the required set of default value converters on their own.

If the `DefaultTypeConverter` does not fit the user's aims, it is easy to define custom type converters, possibly reusing specific ones from the library.

### Acknowledgements
The patch follows the implementation suggested by the [PyAthena](https://github.com/laughingman7743/PyAthena) driver very closely, see [PyAthena/converter.py](https://github.com/laughingman7743/PyAthena/blob/master/pyathena/converter.py). Thank you!

### Synopsis
```python
>>> cursor = connection.cursor(converter=Cursor.get_default_converter())
>>> cursor.execute(stmt)
>>> cursor.fetchone()
['foo', IPv4Address('10.10.10.1'), datetime.datetime(2022, 7, 18, 18, 10, 36, 758000)]
```

### Thoughts
In the form of an early review, please let me know if you endorse this approach and which adjustments you would like to see. 

Personally, I would add more test cases to `test_cursor.py`, in order to exercise the machinery in more detail and to increase code coverage, and probably make the functionality more present in the documentation.

Also, as @mfussenegger mentioned at https://github.com/crate/crate-python/pull/437/files#r930822457, it might make sense to also add the important aspect of converting elements within `ARRAY` types to this patch already.

With kind regards,
Andreas.

---

### Backlog
- [x] Make the functionality more present in the documentation. See 1acc44e1.
- [x] Setting the timezone for datetime objects. See #445. /cc @mfussenegger
- [ ] Decide if the type converter for the `BIT` data type should be included into the `DefaultTypeConverter`.
  https://github.com/crate/crate-python/blob/33ecbf36a3b50606f5138ce9dcee6b00b81996c0/src/crate/client/doctests/cursor.txt#L347

### CODE REVIEW

File: src/crate/client/cursor.py [hunk 1]
```diff
@@ -210,3 +218,23 @@ def duration(self):
...
+        Iterate rows, apply type converters, and generate converted rows.
+        """
+        assert "col_types" in self._result, \
+            "Unable to apply type conversion without `col_types` information"
+        type_id_list = self._result["col_types"]
+        for row in self._result["rows"]:
+            yield [
+                self._converter.convert(type_id, value)
+                for type_id, value in zip(type_id_list, row)
+            ]
```
Comment (amotl) [inline src/crate/client/cursor.py:233]: Hi @matriv,

at https://github.com/crate/crate-python/pull/437#discussion_r924614148, you asked:

> Can we do [this] more elegant?

Maybe this implementation fits your eyes better.

With kind regards,
Andreas.

File: src/crate/client/doctests/cursor.txt [hunk 2]
```diff
@@ -296,6 +304,69 @@ closed connection an ``ProgrammingError`` exception will be raised::
...
+
+CrateDB's ``BIT`` data type has the numeric identifier ``25``. To create a
+simple converter for converging CrateDB's ``BIT`` type to Python's ``int``
+type::
+
+    >>> from crate.client.converter import Converter
+
+    >>> converter = Converter()
+    >>> converter.set(25, lambda value: int(value[2:-1], 2))
+    >>> cursor = connection.cursor(converter=converter)
```
Comment (amotl) [inline src/crate/client/doctests/cursor.txt:348]: @matriv: Simple converters can easily be implemented with lambdas.

**Comment by seut** [comment]
👍 I really like this approach and would definitely prefer it over #437.

File: src/crate/client/converter.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+def _to_ipaddress(value: Optional[str]) -> Optional[Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+    """
+    https://docs.python.org/3/library/ipaddress.html
+    """
+    if value is None:
+        return None
+    return ipaddress.ip_address(value)
+
+
+def _to_datetime(value: Optional[str]) -> Optional[datetime]:
```
Comment (mfussenegger) [inline src/crate/client/converter.py:41]: type annotation is wrong here. A `str` would result in an error with `utcfromtimestamp(value / 1e3)`

File: src/crate/client/converter.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    return datetime.utcfromtimestamp(value / 1e3)
+
+
+def _to_default(value: Optional[str]) -> Optional[str]:
+    return value
+
+
+# Mapping for numeric data type identifiers defined by the CrateDB HTTP interface.
+# https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#column-types
+_DEFAULT_CONVERTERS: Dict[int, Callable[[Optional[int]], Optional[Any]]] = {
```
Comment (mfussenegger) [inline src/crate/client/converter.py:56]: This type annotation results in mypy errors because `_to_ipaddress` and `_to_datetime` have different signatures

File: src/crate/client/converter.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        return self.mappings.get(type_, self._default)
+
+    def set(self, type_: int, converter: Callable[[Optional[int]], Optional[Any]]) -> None:
+        self.mappings[type_] = converter
+
+    def remove(self, type_: int) -> None:
+        self.mappings.pop(type_, None)
+
+    def update(self, mappings: Dict[int, Callable[[Optional[int]], Optional[Any]]]) -> None:
+        self.mappings.update(mappings)
```
Comment (mfussenegger) [inline src/crate/client/converter.py:85]: Not sure if we want to make this mutable?
I also wonder if we should introduce an `Enum` or similar to give names to the type ids?

File: src/crate/client/doctests/cursor.txt [hunk 6]
```diff
@@ -296,6 +304,69 @@ closed connection an ``ProgrammingError`` exception will be raised::
...
+
+    >>> connection.client.set_next_response({
+    ...     "col_types": [4, 5, 11],
+    ...     "rows":[ [ "foo", "10.10.10.1", 1658167836758 ] ],
+    ...     "cols":[ "name", "address", "timestamp" ],
+    ...     "rowcount":1,
+    ...     "duration":123
+    ... })
+
+    >>> cursor.execute('')
```
Comment (mfussenegger) [inline src/crate/client/doctests/cursor.txt:327]: I think this would confuse users because this is a mocked cursor and `client.set_next_response` isn't something that they can use.

Either we need to point this out explicitly, hide it or use real data and don't mock it.

**Review by mfussenegger — COMMENTED** [review]
Left a couple of comments regarding some of the implementation details.


I'm also not sure about the level of separation. E.g. the converter application could probably be separated from the cursor without much loss of API/UI convenience. But no strong opinion here. 

On the other hand, I think something common like setting the timezone for datetime objects could be made even easier for users.

**Comment by mfussenegger** [comment]
There is btw. also https://peps.python.org/pep-0249/#type-objects in the DB API definition - I didn't have a closer look at it, but we should double check to see if it affects what the PR would be doing.

**Comment by amotl** [comment]
Thank you for endorsing this patch and for your valuable suggestions.

File: src/crate/client/cursor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/cursor.py]: Hi. It was just by accident. I just made the method private.

File: src/crate/client/converter.py (hunk 3 above)
Comment (amotl) [reply src/crate/client/converter.py]: Oh, right. Thank for spotting it. After making some adjustments in this area, `mypy src/crate/client/converter.py` doesn't complain anymore now.

File: src/crate/client/converter.py (hunk 4 above)
Comment (amotl) [reply src/crate/client/converter.py]: Fixed. Thank you!

File: src/crate/client/converter.py (hunk 5 above)
Comment (amotl) [reply src/crate/client/converter.py]: > Not sure if we want to make this mutable?

I also wasn't sure, but liked the flexibility in the end. See also my other comment below.

> I also wonder if we should introduce an `Enum` or similar to give names to the type ids?

Good idea. I've introduced an `Enum` with 00dbb42380.

File: src/crate/client/doctests/cursor.txt (hunk 6 above)
Comment (amotl) [reply src/crate/client/doctests/cursor.txt:327]: At the beginning of the file `cursor.txt`, there is a corresponding note about this detail.

> Hardcode the next response of the mocked connection client, so we won't need a sql statement
to execute.

After that, the `set_next_response()` method is already used five times within `cursor.txt`, so I didn't hesitate to continue implementing the test cases this way.

**Comment by amotl** [comment]
Hi again,

other than the inline comments, let me address further questions here.

> Personally, I would add more test cases to `test_cursor.py`, in order to exercise the machinery in more detail and to increase code coverage.

Done, mostly with 141054780f.

> It might make sense to also add the important aspect of converting elements within ARRAY types to this patch already.

Done with b21c38e99.

> I'm also not sure about the level of separation. E.g. the converter application could probably be separated from the cursor without much loss of API/UI convenience. But no strong opinion here.

I also thought back and forth about this detail. In the end, I followed the path outlined by https://github.com/laughingman7743/PyAthena very closely, where I originally discovered this way of doing it. See [PyAthena/converter.py](https://github.com/laughingman7743/PyAthena/blob/master/pyathena/converter.py).

> On the other hand, I think something common like setting the timezone for datetime objects could be made even easier for users.

Indeed, that would be sweet. I will think about how we could add this feature.

> There is btw. also https://peps.python.org/pep-0249/#type-objects in the DB API definition - I didn't have a closer look at it, but we should double check to see if it affects what the PR would be doing.

I also found that section in the DBAPI spec, but I think it is exclusively about **input** parameter conversion:

> Many databases need to have the input in a particular format for binding to an operation’s input parameters. For example, if an input is destined for a DATE column, then it must be bound to the database in a particular string format. 

On the other hand, the current implementation is exclusively about **output** data conversion. That being said, the converter functions itself can well be reused for that use case, but otherwise I think it is a different area, and to be addressed within a different patch.

With kind regards,
Andreas.

---

### Backlog
- Make the functionality more present in the documentation.
- Setting the timezone for datetime objects. /cc @mfussenegger 
- Think about an option to enable the `DefaultTypeConverter` using an SQLAlchemy DBURI connection parameter, because in such scenarios, the implementer would not be able to touch any code at all. This is in order to close the loop where @Aymaru was probably coming from, see #395 and #437.

File: src/crate/client/converter.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    return ipaddress.ip_address(value)
+
+
+def _to_datetime(value: Optional[float]) -> Optional[datetime]:
+    """
+    https://docs.python.org/3/library/datetime.html
+    """
+    if value is None:
+        return None
+    return datetime.utcfromtimestamp(value / 1e3)
```
Comment (amotl) [inline src/crate/client/converter.py:50]: Dear @mfussenegger, @matriv, and @seut,

regarding @mfussenegger's proposal to let the user optionally specify a time zone, and make the machinery return a timezone-aware datetime object, I see two options here.

Let's assume the user defines a custom timezone object like
```python
>>> import datetime
>>> tz_mst = datetime.timezone(datetime.timedelta(hours=7), name="MST")
```

and a timestamp in Epoch like
```python
# Fri, 16 Sep 2022 11:09:16 GMT
timestamp = 1663326556
```

On this spot, which variant would be semantically the right one? Note that [`datetime.utcfromtimestamp`](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) does not accept the `tz` keyword argument, it will always return a naive datetime object with `tzinfo=None`.

1. We can either use [`datetime.fromtimestamp`](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp), which **does** accept the `tz` keyword argument.
   ```python
   >>> datetime.datetime.fromtimestamp(timestamp, tz=tz_mst)
   datetime.datetime(2022, 9, 16, 18, 9, 16, tzinfo=datetime.timezone(datetime.timedelta(seconds=25200), 'MST'))
   ```

2. Or we can replace the `tzinfo` on the naive datetime object returned by `utcfromtimestamp`, making it timezone-aware.
   ```python
   >>> datetime.datetime.utcfromtimestamp(timestamp).replace(tzinfo=tz_mst)
   datetime.datetime(2022, 9, 16, 11, 9, 16, tzinfo=datetime.timezone(datetime.timedelta(seconds=25200), 'MST'))
   ```

I think option 1. is the right choice. So the implementation would look like something along the lines of:
```python

def set_timezone(self, tz):
  if not isinstance(tz, datetime.timezone):
    raise TypeError(f"Timezone object has wrong type: {tz}")
  self.timezone = tz

def _to_datetime(self, value):
  """
  Convert CrateDB's `TIMESTAMP` column to the native Python `datetime` object.
  When a timezone is given, return a timezone-aware object.
  Otherwise, return a naive object which is meant to be the time in the UTC time zone.
  """
  if self.timezone is not None:
    return datetime.fromtimestamp(value / 1e3, tz=self.timezone)
  else:
    return datetime.utcfromtimestamp(value / 1e3)
```

What do you think?

With kind regards,
Andreas.

Comment (amotl) [reply src/crate/client/converter.py:50]: Another implementation could look like this, it is **always** using `datetime.fromtimestamp(value / 1e3, tz=self.timezone)`, where `self.timezone` would be `datetime.timezone.utc` by default, i.e. it will always return timezone-aware datetime objects.

While definitively easier and more straight-forward, it could be less performant when **not** aiming to use timezone-aware datetime objects and stay in "naive" land **on purpose**.

```python
def __init__(self):
  # Return datetype objects in UTC by default.
  self.timezone = datetime.timezone.utc

def set_timezone(self, tz):
  """
  Specify a time zone if you want to have your datetime objects in a time zone other than UTC.
  """
  if not isinstance(tz, datetime.timezone):
    raise TypeError(f"Timezone object has wrong type: {tz}")
  self.timezone = tz

def _to_datetime(self, value):
  """
  Convert CrateDB's `TIMESTAMP` column to a timezone-aware native Python `datetime` object.
  When no timezone is given, UTC is used.
  """
  return datetime.fromtimestamp(value / 1e3, tz=self.timezone)
```

Comment (amotl) [reply src/crate/client/converter.py:50]: > definitively easier and more straight-forward

When thinking about it once more, I think I would prefer the implementation variant outlined in my latest post. In this way, it is always consistent that timezone-aware datetime objects are returned from the data converter machinery.

> it could be less performant when not aiming to use timezone-aware datetime objects and stay in "naive" land on purpose.

I think when users are aiming for raw speed, they would completely opt-out of the data type converter machinery at all.

Comment (amotl) [reply src/crate/client/converter.py:50]: Hi again. Talk is cheap. I've converged a few more thoughts about this into a proposed patch at #445, completely separating the topic of timezone-aware `datetime` handling from this one.

File: src/crate/client/converter.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class Converter:
+    def __init__(
+        self,
+        mappings: Dict[int, Callable[[Optional[InputVal]], Optional[Any]]] = None,
+        default: Callable[[Optional[InputVal]], Optional[Any]] = _to_default,
+    ) -> None:
+        self._mappings = mappings or {}
+        self._default = default
```
Comment (mfussenegger) [inline src/crate/client/converter.py:102]: Could promote usage of the enum further and use it in `mappings` as key instead of the `int` value.

That would implicitly document what `int` values are acceptable.

File: src/crate/client/converter.py (hunk 5 above)
Comment (mfussenegger) [reply src/crate/client/converter.py]: > I also wasn't sure, but liked the flexibility in the end to make the user able to add and remove converters as they like.

I'd keep it immutable for now. It reduces the API surface and makes it easier to reason about the mappings.
And _if_ a legitimate use-case pops up, it's easy to change later. On the other hand, changing a mutable mapping to immutable later on would be a breaking change.

File: src/crate/client/doctests/cursor.txt (hunk 6 above)
Comment (mfussenegger) [reply src/crate/client/doctests/cursor.txt:327]: Is there a reason we can't/shouldn't hide it for the rendered docs?

File: src/crate/client/converter.py (hunk 5 above)
Comment (amotl) [reply src/crate/client/converter.py]: Are you fine with removing mutability by 299b7713ad?

File: src/crate/client/doctests/cursor.txt (hunk 6 above)
Comment (amotl) [reply src/crate/client/doctests/cursor.txt:327]: As far as I can see, `cursor.txt` is not part of the rendered docs at https://crate.io/docs/python/, it is only used as a doctest.

File: src/crate/client/converter.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    return datetime.utcfromtimestamp(value / 1e3)
+
+
+def _to_default(value: Optional[InputVal]) -> Optional[Any]:
+    return value
+
+
+# Symbolic aliases for the numeric data type identifiers defined by the CrateDB HTTP interface.
+# https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#column-types
+class CrateDatatypeIdentifier(Enum):
```
[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/converter.py]: Thanks. Renamed with 134d634.

File: src/crate/client/converter.py (hunk 8 above)
Comment (amotl) [reply src/crate/client/converter.py]: Thank you for bringing this up.

### Problem
Without looking further into the details and the state of the onion, I was wary of using an Enum-type as key, because I found different resources like those referenced below, which indicate possible performance losses. Especially because the converter sits in a busy place of the driver, I decided to not use it this way and use an `int` as key instead.

### Thoughts
If we still want to use an Enum-type as a key on the mapping data structure itself, I can think of two options, but I will be happy to hear if you see others.
- Use in-memory `@memoize`-style caching on the lookup path, if that would work around the Enum-lookup-is-slow problem.
- Look into using the `fastenum` library.

### References

TLDR; Apparently, Python 3.4's first Enum implementation was pretty slow. While it got better with subsequent releases, the `fastenum` library still claims to be up to 3.5x faster in benchmarks on recent versions of Python.

- https://quantlane.com/blog/fastenum/, https://gitlab.com/quantlane/libs/fastenum
- https://pypi.org/project/fast-enum/
- https://bugs.python.org/issue39102

Comment (mfussenegger) [reply src/crate/client/converter.py]: Is this relevant for how we use it?

What we should do independent of this is to  move the dict lookup out of the row loop. The columns/types are the same across all rows. We can retrieve the converter for all columns once and then process all rows with them.

Then you've 1 enum attribute access and 1 dict lookup *per* column. That shouldn't be a performance issue.

File: src/crate/client/tests.py [hunk 10]
```diff
@@ -335,6 +312,10 @@ def _try_execute(cursor, stmt):
     try:
         cursor.execute(stmt)
     except Exception:
+        # FIXME: Why does this croak on statements like ``DROP TABLE cities``?
+        # Note: When needing to debug the test environment, you may want to
+        #       enable this logger statement.
+        # log.exception("Executing SQL statement failed")
```
Comment (amotl) [inline src/crate/client/tests.py:318]: ### About
Tests: Add inline comment about how to debug the test environment.

### Background story
It was needed because the tests suddenly started failing mysteriously on my machine.

Out of the blue, the test suite started croaking like `TABLE 'locations' already exists`, only to find that CrateDB would not accept any write operations after finding that my disk usage was 96%, while I had still 60G of free disk space.

Unfortunately, I did not discover this easily, because the test suite swallowed the exception on this spot. When enabling the `log.exception` statement, the error response from CrateDB became immediately visible.

File: src/crate/client/converter.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    CHARACTER = 27
+    ARRAY = 100
+
+
+# Map data type identifier to converter function.
+_DEFAULT_CONVERTERS: Dict[DataType, ConverterFunction] = {
+    DataType.IP: _to_ipaddress,
+    DataType.TIMESTAMP_WITH_TZ: _to_datetime,
+    DataType.TIMESTAMP_WITHOUT_TZ: _to_datetime,
+}
```
Comment (amotl) [inline src/crate/client/converter.py:97]: Not sure if the `DefaultTypeConverter` should apply the `_to_datetime` function to _both_ kinds of `TIMESTAMP` types?

File: src/crate/client/doctests/cursor.txt [hunk 12]
```diff
@@ -296,6 +304,69 @@ closed connection an ``ProgrammingError`` exception will be raised::
...
+    ...     "rows":[ [ "B'0110'" ] ],
+    ...     "cols":[ "value" ],
+    ...     "rowcount":1,
+    ...     "duration":123
+    ... })
+
+    >>> cursor.execute('')
+
+    >>> cursor.fetchone()
+    [6]
```
Comment (amotl) [inline src/crate/client/doctests/cursor.txt:364]: Do you think the type converter for the `BIT` data type should be included into the `DefaultTypeConverter`, like the converters for `IP` and `TIMESTAMP`? I am talking about this one:

https://github.com/crate/crate-python/blob/33ecbf36a3b50606f5138ce9dcee6b00b81996c0/src/crate/client/doctests/cursor.txt#L347

File: src/crate/client/doctests/cursor.txt (hunk 6 above)
Comment (amotl) [reply src/crate/client/doctests/cursor.txt:327]: I've created #448 to have a separate discussion about the topic how to improve/adjust doctests vs. rendered documentation.

File: src/crate/client/converter.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        self,
+        mappings: Optional[Dict[DataType, ConverterFunction]] = None,
+        default: ConverterFunction = _to_default,
+    ) -> None:
+        self._mappings = mappings or {}
+        self._default = default
+
+    @property
+    def mappings(self) -> Dict[DataType, ConverterFunction]:
+        return self._mappings
```
Comment (mfussenegger) [inline src/crate/client/converter.py:111]:
```suggestion
```

File: src/crate/client/converter.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+import ipaddress
+from copy import deepcopy
+from datetime import datetime
+from enum import Enum
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+InputVal = Any
+ConverterFunction = Callable[[Optional[InputVal]], Optional[Any]]
+ConverterDefinition = Union[ConverterFunction, Tuple[ConverterFunction, "ConverterDefinition"]]
+ColTypesDefinition = Union[int, List[Union[int, "ColTypesDefinition"]]]
```
Comment (mfussenegger) [inline src/crate/client/converter.py:35]:
```suggestion
ConverterFunction = Callable[[Optional[Any]], Optional[Any]]
ConverterDefinition = Union[ConverterFunction, Tuple[ConverterFunction, "ConverterDefinition"]]
ColTypesDefinition = Union[int, List[Union[int, "ColTypesDefinition"]]]
```

No need for custom vocabulary for `Any`?

File: src/crate/client/doctests/cursor.txt (hunk 6 above)
Comment (amotl) [reply src/crate/client/doctests/cursor.txt:327]: I totally agree with you. I think your guidance should be implemented seperately, being tracked by #448. If you want to see it sooner than later, than I will be happy to tackle it before coming back to this patch.

Comment (mfussenegger) [reply src/crate/client/doctests/cursor.txt:327]: Sounds good to handle this separately in a follow up

File: src/crate/client/converter.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+    @property
+    def mappings(self) -> Dict[DataType, ConverterFunction]:
+        return self._mappings
+
+    def get(self, type_: DataType) -> ConverterFunction:
+        return self.mappings.get(type_, self._default)
+
+    def set(self, type_: DataType, converter: ConverterFunction) -> None:
+        self.mappings[type_] = converter
```
[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/converter.py]: I've just added 14a355daa, to follow your suggestion.

File: src/crate/client/doctests/cursor.txt (hunk 12 above)
Comment (mfussenegger) [reply src/crate/client/doctests/cursor.txt]: I'd leave it out for now. Easy enough to add later if requested by users

File: src/crate/client/connection.py [hunk 16]
```diff
@@ -123,12 +130,16 @@ def __init__(self,
         self.lowest_server_version = self._lowest_server_version()
         self._closed = False
 
-    def cursor(self):
+    def cursor(self, cursor=None, **kwargs) -> Cursor:
```
Comment (mfussenegger) [inline src/crate/client/connection.py:133]: What's the `cursor=None` for?

File: src/crate/client/cursor.py [hunk 17]
```diff
@@ -210,3 +221,31 @@ def duration(self):
...
+                convert(value)
+                for convert, value in zip(converters, row)
+            ]
+
+    @staticmethod
+    def get_default_converter(mappings: Optional[ConverterMapping] = None) -> Converter:
+        """
+        Return the standard converter instance.
+        """
+        return DefaultTypeConverter(more_mappings=mappings)
```
Comment (mfussenegger) [inline src/crate/client/cursor.py:251]: Is this needed? It's just wrapping a constructor which could be called directly?

**Review by mfussenegger — COMMENTED** [review]
Left two more comments, otherwise looks good to me

File: src/crate/client/connection.py (hunk 16 above)
Comment (amotl) [reply src/crate/client/connection.py:133]: It probably slipped in from the role model implementation of PyAthena. I think it can be removed. Thanks for spotting it.

```
def cursor(self, cursor: Optional[Type[BaseCursor]] = None, **kwargs) -> BaseCursor:
```

-- https://github.com/laughingman7743/PyAthena/blob/2f2c0e11487426bd382f480a8468e9bc78c42156/pyathena/connection.py#L231

File: src/crate/client/cursor.py (hunk 17 above)
Comment (amotl) [reply src/crate/client/cursor.py:251]: It looks like `get_default_converter()` is currently never called with any parameters, so I guess it can also be removed. Thanks!

**Review by mfussenegger — APPROVED** [review]
